### PR TITLE
feat: add Core API pytorch mnist tutorial update

### PR DESCRIPTION
**Pull request by ighodgao**
Both types of continuations tested and ready for review

### CODE REVIEW

File: examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py [hunk 1]
```diff
@@ -97,11 +100,23 @@ def test(args, model, device, test_loader, epoch, core_context, steps_completed)
     )
 
 
-# NEW: Define load_state function for restarting model training from existing checkpoint.
-def load_state(checkpoint_directory):
+# NEW: Define load_state function for restarting model training from existing checkpoint. Returns (.pt, int).
+# Also update load_state header to take trial info object as an argument.
+def load_state(checkpoint_directory, info):
```
Comment (rb-determined-ai) [inline examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py:105]: accept a `trial_id: int` here instead of an `info: ClusterInfo`.  No need to pass a huge object just for one integer it contains.

File: examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py [hunk 2]
```diff
@@ -97,11 +100,23 @@ def test(args, model, device, test_loader, epoch, core_context, steps_completed)
...
     checkpoint_directory = pathlib.Path(checkpoint_directory)
-    with checkpoint_directory.joinpath("checkpoint.pt").open("rb") as f:
-        return torch.load(f)
+
+    with checkpoint_directory.joinpath("state").open("r") as f:
+        epochs_completed, ckpt_trial_id = [int(field) for field in f.read().split(",")]
+
+    # NEW: If the saved trial ID is the same as our current trial ID, we are resuming the trial, so we should return the model weights
```
Comment (rb-determined-ai) [inline examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py:111]: isn't this whole function already marked as `# NEW:`?  I think you could omit the `NEW:` portion of this comment if you'd like.

File: examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py [hunk 3]
```diff
@@ -166,7 +181,9 @@ def main(core_context):
     latest_checkpoint = info.latest_checkpoint
     if latest_checkpoint is not None:
         with core_context.checkpoint.restore_path(latest_checkpoint) as path:
-            model = load_state(path)
+            model, epochs_completed = load_state(path, info)
+    else:
+        epochs_completed = 0
```
Comment (rb-determined-ai) [inline examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py:186]: nit: slight preference for reversing the `if` block order:

```python
if latest_checkpoint is None:
  epochs_completed = 0
else:
  with core_context.checkpoint.restore_path(...):
    ...
```

Reason: it's easier to read code if the physical code layout mirrors the execution order.  The way I am proposing it, the first time you run the code you hit the first block, the second time (if you pause/activate) you'll go through the second block.

File: examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py [hunk 4]
```diff
@@ -196,7 +213,9 @@ def main(core_context):
     optimizer = optim.Adadelta(model.parameters(), lr=args.lr)
 
     scheduler = StepLR(optimizer, step_size=1, gamma=args.gamma)
-    for epoch in range(1, args.epochs + 1):
+
+    # NEW: Resume training from epochs_completed. This is useful in the case of pausing and resuming an experiment.
+    for epoch in range(epochs_completed, args.epochs + 1):
 
         steps_completed = epoch * len(train_loader)
```
Comment (rb-determined-ai) [inline examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py:220]: I said this before I think, but it reads awfully strange to set steps_completed before calling `train()`, since at this point you have actually completed zero steps.

File: examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py [hunk 5]
```diff
@@ -196,7 +213,9 @@ def main(core_context):
     optimizer = optim.Adadelta(model.parameters(), lr=args.lr)
 
     scheduler = StepLR(optimizer, step_size=1, gamma=args.gamma)
-    for epoch in range(1, args.epochs + 1):
+
+    # NEW: Resume training from epochs_completed. This is useful in the case of pausing and resuming an experiment.
+    for epoch in range(epochs_completed, args.epochs + 1):
```
Comment (rb-determined-ai) [inline examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py:218]: two things:

1. you have a bug relative to the old code where the old code started at epoch=1 but this code starts at epoch=0
2. it is way more common to do ranges starting from zero, and you can use the name "epoch_idx" to make it more obvious when reading the code that it's a zero-indexed value rather than a 1-indexed value.  In fact, to make it super easy to read, you can even show the relationship between the `epoch_idx` and the `epochs_completed`:

```python
for epoch_idx in range(epochs_completed, args.epochs):
    train(...)
    epochs_completed = epoch_idx + 1
    steps_completed = epochs_completed * len(train_loader)
    test(...)

    ...
    with core_context.checkpoint.store_path(...):
        with path.joinpath("state").open("w") as f:
            f.write(f"{epochs_completed},{info.trial.trial._id}")
```

File: examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py [hunk 6]
```diff
@@ -208,8 +227,12 @@ def main(core_context):
         # NEW: Save checkpoint.
         checkpoint_metadata_dict = {"steps_completed": steps_completed}
 
+        # NEW: Here we are saving multiple files to our checkpoint directory, a model weights file as well as a file that includes other
+        # information about the checkpoint, in this case, epochs completed and the trial ID.
```
Comment (rb-determined-ai) [inline examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py:231]: Break this comment into multiple sentences, or bullet points, or something.  It is not very understandable.

Also I recommend the phrases "model state" and "training loop state" or something to that effect in order to illustrate the difference, but up to you.

File: examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py [hunk 7]
```diff
@@ -97,11 +100,23 @@ def test(args, model, device, test_loader, epoch, core_context, steps_completed)
...
+        epochs_completed, ckpt_trial_id = [int(field) for field in f.read().split(",")]
+
+    # NEW: If the saved trial ID is the same as our current trial ID, we are resuming the trial, so we should return the model weights
+    # and the number of epochs already completed to resume from.
+    if ckpt_trial_id == info.trial.trial_id:
+        with checkpoint_directory.joinpath("checkpoint.pt").open("rb") as saved_model:
+            return torch.load(saved_model), epochs_completed
+    else:
+        # This is a new trial; load the model weight but return the number of epochs completed as 0.
+        with checkpoint_directory.joinpath("checkpoint.pt").open("rb") as saved_model:
```
Comment (rb-determined-ai) [inline examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py:118]: copy/paste error: these should not both be called "saved_model" since they don't both contain models.  If you want to call them both `f` that's fine, but if you give them descriptive names they need to be correct.

File: examples/tutorials/core_api_pytorch_mnist/model_def_distributed.py [hunk 8]
```diff
@@ -234,7 +243,7 @@ def main(core_context):
     optimizer = optim.Adadelta(model.parameters(), lr=hparams["learning_rate"])
     scheduler = StepLR(optimizer, step_size=1, gamma=args.gamma)
 
-    starting_epoch = 0
```
Comment (rb-determined-ai) [inline examples/tutorials/core_api_pytorch_mnist/model_def_distributed.py:237]: this was always a bug, since you have a zero-indexed `epoch` but you calculate `steps_completed = epoch * len(training_loader)`, which will be zero after you train one epoch.

Again, my recommendation about being explicit naming `epoch_idx` and `epochs_completed` helps avoid these off-by-one errors.

Side note, there's no reason for this `starting_epoch` variable, prolly get rid of it.

File: examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py (hunk 4 above)
Comment (ighodgao) [reply examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py:220]: This is the cleanest way to pass steps_completed into test() though

File: examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py (hunk 7 above)
Comment (ighodgao) [reply examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py:118]: By both of these not containing models, you mean in the else case it'd be an empty `checkpoint.pt` file correct?

Comment (rb-determined-ai) [reply examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py:118]: oh snap, I totally misread the code.  Going to reread and re-review this bit.

File: examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py [hunk 9]
```diff
@@ -97,11 +100,23 @@ def test(args, model, device, test_loader, epoch, core_context, steps_completed)
...
+            return torch.load(saved_model), 0
```
Comment (rb-determined-ai) [inline examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py]: There's no need to duplicate your loading code in both if statements like this.

Why not:
```python

with checkpoint_directory.joinpath(...) as f:
    model = torch.load(saved_model)

# If trial ID does not match our current trial ID, we'll ignore epochs
# completed and start training from epoch_idx = 0
if ckpt_trial_id == current_trial_id:
    epochs_completed = 0

return model, epochs_completed
```

File: examples/tutorials/core_api_pytorch_mnist/distributed.yaml [hunk 10]
```diff
@@ -35,4 +35,8 @@ searcher:
   max_length:
     epochs: 20
 resources:
-  slots_per_trial: 4
+  slots_per_trial: 2
+environment:
+  environment_variables:
+    # You may need to modify this to match your network configuration.
+    - NCCL_SOCKET_IFNAME=ens,eth,ib
```
Comment (rb-determined-ai) [inline examples/tutorials/core_api_pytorch_mnist/distributed.yaml:42]: remove this

File: examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py [hunk 11]
```diff
@@ -196,20 +209,34 @@ def main(core_context):
...
 
     scheduler = StepLR(optimizer, step_size=1, gamma=args.gamma)
-    for epoch in range(1, args.epochs + 1):
 
-        steps_completed = epoch * len(train_loader)
-
-        train(args, model, device, train_loader, optimizer, epoch, core_context)
-        test(args, model, device, test_loader, epoch, core_context, steps_completed=steps_completed)
+    # NEW: Resume training from epochs_completed. This is useful in the case of pausing and resuming an experiment.
+    for epoch_idx in range(epochs_completed, args.epochs + 1):
```
Comment (rb-determined-ai) [inline examples/tutorials/core_api_pytorch_mnist/model_def_checkpoints.py:214]: you have an off-by-one here due to the lingering `args.epochs + 1`, even though you are starting at 0 now

File: examples/tutorials/core_api_pytorch_mnist/distributed.yaml [hunk 12]
```diff
@@ -35,4 +35,4 @@ searcher:
   max_length:
     epochs: 20
 resources:
-  slots_per_trial: 4
+  slots_per_trial: 2
```
Comment (rb-determined-ai) [inline examples/tutorials/core_api_pytorch_mnist/distributed.yaml:38]: this change still shouldn't be here

**Review by rb-determined-ai — APPROVED** [review]
great!

**Review by tara-det-ai — APPROVED** [review]
LGTM

File: examples/tutorials/core_api_pytorch_mnist/model_def_distributed.py [hunk 13]
```diff
@@ -234,15 +247,15 @@ def main(core_context):
...
+    epoch_idx = epochs_completed
     last_checkpoint_batch = None
 
     for op in core_context.searcher.operations():
 
-        while epoch < op.length:
-            steps_completed = epoch * len(train_loader)
-            train(args, model, device, train_loader, optimizer, core_context, epoch, op)
+        while epoch_idx < op.length:
+            train(args, model, device, train_loader, optimizer, core_context, epoch_idx, op)
```
Comment (rb-determined-ai) [inline examples/tutorials/core_api_pytorch_mnist/model_def_distributed.py:256]: I noticed looking at the proposed docs that there's still an off-by-one error here: you are calling `train(epoch=epoch_idx)` basically but you are calculating steps_completed inside of `train()` based on `(epoch - 1)`, which will dip down to `-1` at `epoch_idx=0`.

I suggest finding all instances of `epoch` in your examples and being consistent as to what should be `epoch_idx` and what should be `epochs_completed`.  In the case of `train()` for instance, I think passing `epoch_idx` makes the most sense, and since `epochs_completed` isn't really "completed" until after `train()` is actually done.

File: examples/tutorials/core_api_pytorch_mnist/distributed.yaml (hunk 12 above)
Comment (tara-det-ai) [reply examples/tutorials/core_api_pytorch_mnist/distributed.yaml]: resolved

File: examples/tutorials/core_api_pytorch_mnist/model_def_distributed.py (hunk 13 above)
Comment (tara-det-ai) [reply examples/tutorials/core_api_pytorch_mnist/model_def_distributed.py:256]: resolved

File: examples/tutorials/core_api_pytorch_mnist/model_def_metrics.py [hunk 14]
```diff
@@ -193,7 +193,7 @@ def main(core_context):
     optimizer = optim.Adadelta(model.parameters(), lr=args.lr)
 
     scheduler = StepLR(optimizer, step_size=1, gamma=args.gamma)
-    for epoch in range(1, args.epochs + 1):
+    for epoch in range(0, args.epochs):
```
Comment (rb-determined-ai) [inline examples/tutorials/core_api_pytorch_mnist/model_def_metrics.py:196]: I recommend changing from `epoch` to `epoch_idx` here, for clarity.  Also, up to you, but you might consider leaving a `# NEW` comment explaining the reason why we want the extra clarity of `epoch_idx`.

File: examples/tutorials/core_api_pytorch_mnist/model_def_adaptive.py [hunk 15]
```diff
@@ -108,10 +109,21 @@ def test(args, model, device, test_loader, core_context, steps_completed, op) ->
     return test_loss
 
 
-def load_state(checkpoint_directory):
+def load_state(checkpoint_directory, trial_id):
+
```
Comment (rb-determined-ai) [inline examples/tutorials/core_api_pytorch_mnist/model_def_adaptive.py:113]: unnecessary added empty line here, doesn't match `model_def_checkpoints.py`

File: examples/tutorials/core_api_pytorch_mnist/model_def_adaptive.py [hunk 16]
```diff
@@ -211,18 +225,19 @@ def main(core_context):
...
     for op in core_context.searcher.operations():
 
         # NEW: Use a while loop for easier accounting of absolute lengths.
-        while epoch < op.length:
-            steps_completed = epoch * len(train_loader)
+        while epoch_idx < op.length:
 
             # NEW: Pass op into train() and test().
-            train(args, model, device, train_loader, optimizer, core_context, epoch, op)
+            train(args, model, device, train_loader, optimizer, core_context, epoch_idx, op)
```
Comment (rb-determined-ai) [inline examples/tutorials/core_api_pytorch_mnist/model_def_adaptive.py:238]: It's confusing in an example to pass `epoch_idx` into `train()`, but inside `train()` it's called `epoch` instead.  I really think you should update all instances of `epoch` to reflect the exact meaning (`epoch_idx` or `epochs_completed`), so quantities are easier to track across function calls.

**Review by rb-determined-ai — APPROVED** [review]
The remaining issues are much smaller, feel free to land after this.

I don't plan on re-reviewing, it looks pretty good.

File: examples/tutorials/core_api_pytorch_mnist/model_def_metrics.py [hunk 17]
```diff
@@ -69,7 +69,7 @@ def train(args, model, device, train_loader, optimizer, epoch, core_context):
             # NEW: Report epoch-based training metrics to Determined master via core_context.
             # Index by (batch_idx + 1) * (epoch-1) * len(train_loader) to continuously plot loss on one graph for consecutive epochs.
             core_context.train.report_training_metrics(
-                steps_completed=(batch_idx + 1) + (epoch - 1) * len(train_loader),
+                steps_completed=(batch_idx + 1) + epoch * len(train_loader),
                 metrics={"train_loss": loss.item(), "epoch": epoch},
```
[the remark this answers is not in the document]
Comment (ighodgao) [reply examples/tutorials/core_api_pytorch_mnist/model_def_metrics.py:73]: because I changed epoch indexing to start from 0 rather than 1 in the train loop, I also changed this part to make sure the metric value did not change

**Comment by rb-determined-ai** [comment]
I'm overriding the CI system here because the only failure is a generic proxy test, and that is a known broken test (and @tara-det-ai has been blocked on this for days)
